app/controllers/composer_controller.rb

```diff
@@ -0,0 +1,166 @@
+        .pluck(:user_id)
+        .to_set
+    else
+      Set.new
+    end
+
+    topic_allowed_user_ids = if allowed_names.present?
+      User
+        .where(username_lower: allowed_names.map(&:downcase))
+        .pluck(:id)
+        .to_set
+    elsif topic&.private_message?
+      TopicAllowedUser
+        .where(topic: topic)
+        .pluck(:user_id)
+        .to_set
+    end
+
+    topic_allowed_group_ids = if allowed_names.present?
+      Group
+        .messageable(current_user)
+        .where(name: allowed_names)
+        .pluck(:id)
+        .to_set
+    elsif topic&.private_message?
+      TopicAllowedGroup
+        .where(topic: topic)
+        .pluck(:group_id)
+        .to_set
+    end
+
+    names.each do |username|
+      user = users[username]
+      next if user.blank?
+
+      reason = if topic && !user.guardian.can_see?(topic)
+        topic.private_message? ? :private : :category
+      elsif allowed_names.present? && !is_user_allowed?(user, topic_allowed_user_ids, topic_allowed_group_ids)
+        # This would normally be handled by the previous if, but that does not work for new private messages.
+        :private
+      elsif topic_muted_by.include?(user.id)
+        :muted_topic
+      elsif topic&.private_message? && !is_user_allowed?(user, topic_allowed_user_ids, topic_allowed_group_ids)
+        # Admins can see the topic, but they will not be mentioned if they were not invited.
+        :not_allowed
+      end
+
+      # Regular users can see only basic information why the users cannot see the topic.
+      reason = nil if !guardian.is_staff? && reason != :private && reason != :category
+
+      user_reasons[username] = reason if reason.present?
+    end
+
+    names.each do |name|
+      group = groups[name]
+      next if group.blank?
+
+      group_reasons[name] = if !mentionable_group_ids.include?(group.id)
+        :not_mentionable
+      elsif (topic&.private_message? || allowed_names.present?) && !topic_allowed_group_ids.include?(group.id)
+        :not_allowed
+      end
+    end
+
+    if topic && names.include?(SiteSetting.here_mention) && guardian.can_mention_here?
+      here_count = PostAlerter.new.expand_here_mention(topic.first_post, exclude_ids: [current_user.id]).size
+    end
+
+    serialized_groups = groups.values.map do |group|
+      serialized_group = { user_count: group.user_count }
+
+      if group_reasons[group.name] == :not_allowed &&
+          members_visible_group_ids.include?(group.id) &&
+          (topic&.private_message? || allowed_names.present?)
+
+        # Find users that are notified already because they have been invited
+        # directly or via a group
+        notified_count = GroupUser
+          # invited directly
+          .where(user_id: topic_allowed_user_ids)
+          .or(
+            # invited via a group
+            GroupUser.where(
+              user_id: GroupUser.where(group_id: topic_allowed_group_ids).select(:user_id)
```

Comment:
I'm a little confused here with `group_id: topic_allowed_group_ids` because the query does a `.where(group_id: group.id)` below so there is no need for us to look at any other groups in this filter here except for `group.id`?

Reply:
This query returns the number of users that are member of the target groups AND (have been invited directly OR a member of the invited groups). I wanted a single query with a UNION that fetches all the data:

* the users invited directly - `topic_allowed_user_ids` that are members of `group.id`
* the users invited indirectly - users that are members of `topic_allowed_group_ids` that are members of `group.id` too

Since the "members of `group.id`" part is duplicated I put it after the `.or()` call, but the other parts should match perfectly.

Reply:
Thank you for explaining :+1: I have a much better understanding of what the query is trying to achieve now.